### ISSUE TITLE
Custom trained object detection model not working

### ISSUE THREAD

**Issue by RockZombie4** [ultralytics/yolov5#8528]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Training, Detection

### Bug

I have recently trained a custom object detection model for YOLOv5 with makesense.ai .

I exported as a label file, and I put 76 images in the proper training directory, and 8 images in the proper validation directory.
I just need it to detect my face as a test but whenever I run `detect.py` with the custom weights generated by `train.py`, it never detects. There is no error message being output in the logs, but for some reason it just doesn't show any detection box around my face. 

Here is my yolov5 custom model:
[datasets.zip](https://github.com/ultralytics/yolov5/files/9075344/datasets.zip)

I used this command to attempt to detect an image with my face:
`python detect.py --data data/coco128.yaml --weights runs/train/exp10/weights/best.pt --source ~/Pictures/Webcam/2022-07-05-064922.jpg`

I used this command to train the dataset:
`python train.py --data data/coco128.yaml --epochs 1 --imgsz 76`

Please help me!

### Environment

- YOLO - YOLOv5
- OS - Raspbian OS 64-bit
- Python 3.7.10

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@RockZombie4 I doubt your training will work with so few images. If you only have 76 images you'll probably want to train very many epochs, i.e. >>1000 to get enough iterations to produce results.

Alternatively simply lower your --conf-thresh. 

Obviously at some point as confidence thresholds trends to zero boxes will begin to appear.

**Comment by RockZombie4** [comment]
> `python train.py --data data/coco128.yaml --epochs 1 --imgsz 76`

How do I lower the `--conf-thresh`?

**Comment by glenn-jocher** [comment]
@RockZombie4 https://github.com/ultralytics/yolov5/blob/526e650553819dbff67897b9c752c4072e989823/detect.py#L221

**Comment by glenn-jocher** [comment]
@RockZombie4 but I have no idea why you would expect results from 1 epoch of training. This is useless. See Train Custom Data for directions on training and Tips for Improving Results tutorials:

## YOLOv5 Tutorials

* [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)&nbsp; 🚀 RECOMMENDED
* [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)&nbsp; ☘️  RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Roboflow for Datasets, Labeling, and Active Learning](https://docs.ultralytics.com/yolov5/tutorials/roboflow_datasets_integration)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
* [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)&nbsp; ⭐ NEW
* [TFLite, ONNX, CoreML, TensorRT Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) 🚀
* [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
* [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
* [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
* [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
* [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)&nbsp; ⭐ NEW
* [Architecture Summary](https://docs.ultralytics.com/yolov5/tutorials/architecture_description)&nbsp; ⭐ NEW

Good luck 🍀 and let us know if you have any other questions!

**Comment by RockZombie4** [comment]
I tried training with 1000 epochs, set the image size to 96, and lowered the `--conf-thresh` still nothing!

Do  I train with a pretrained weight?
What do I do?

**Comment by glenn-jocher** [comment]
@RockZombie4 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by HusseinA190** [comment]
> @RockZombie4 but I have no idea why you would expect results from 1 epoch of training. This is useless. See Train Custom Data for directions on training and Tips for Improving Results tutorials:

#but i have trained my model on a large data like 5,000 images and trained them on 100 epoch and the weights file dosen't even detect anything and the accuracy was 86%

**Comment by greeshmathilakan369** [comment]
Bug
I have  trained a custom object detection model for YOLOv5.  There is no error message being output in the logs, but for some reason it just doesn't show any detection box around my face.

image 1/1: 720x1280 (no detections)
Speed: 1026.1ms pre-process, 167.8ms inference, 1.0ms NMS per image at shape (1, 3, 384, 640)

**Comment by glenn-jocher** [comment]
@RockZombie4 👋 Hello! Thanks for reaching out. 

If you are getting an accuracy of 86% but no detections are being made, there could be several reasons why this is happening. I suggest you do the following:

1. **Check that your dataset is labeled correctly.** Ensure that all of your images have accurate annotations and bounding boxes around the target objects you want to detect. Ensure that the annotations are in `xywh` format and are saved in a compatible format, such as YOLO format or COCO format.

2. **Try testing your model on different images.** There is always a possibility that your dataset may not include enough variety of images for the model to generalize well. Therefore, testing on various other images would enable you to check the model's robustness.

3. **Change the Confidence Threshold.** Try using different values of `--conf-thres` to see if it resolves the issues with lower detection rates. By default, YOLOv5 uses a confidence threshold of `0.25`. A higher threshold value will produce less detections, but higher precision.

4. **Try training with more epochs.** Try training the model for more epochs and see if it improves the accuracy of detection. 

5. **Data Augmentation Changes.** Try varying the data augmentation methods, such as flipping, scaling, translation, to get enough augmentations for better model generalization.
 
If none of these solutions work, then provide detail information about your problem as well as the information in the log files so that we can diagnose the issue further.

**Comment by harshrocks** [comment]
> @RockZombie4 I doubt your training will work with so few images. If you only have 76 images you'll probably want to train very many epochs, i.e. >>1000 to get enough iterations to produce results.
> 
> Alternatively simply lower your --conf-thresh.
> 
> Obviously at some point as confidence thresholds trends to zero boxes will begin to appear.

Lowering the --conf value did the work thank you so much! You're an angel!😊😊😇

**Comment by glenn-jocher** [comment]
@harshrocks glad to hear that lowering the `--conf` value resolved the issue for you! It's always a good idea to experiment with different threshold values to fine-tune the detection output according to your requirements. If you have any further questions or need assistance with anything else, please feel free to ask. Happy detecting! 😊😇

**Comment by alienCarpet14** [comment]
> > @RockZombie4 but I have no idea why you would expect results from 1 epoch of training. This is useless. See Train Custom Data for directions on training and Tips for Improving Results tutorials:
> 
> #but i have trained my model on a large data like 5,000 images and trained them on 100 epoch and the weights file dosen't even detect anything and the accuracy was 86%

Had similar issue here. Trained my dataset it was accurate on validation set but my dumb self predicted with targeting to classes=[9] because I was reusing code that detected Traffic Lights with yolov8n where Traffic Light was on the 9th index of classes that it detects. And I trained my model to detect only one class so I had to change it to classes=[0].
After reading this forum I had come to this idea. Was going mad last 2 days about this. Funny how all bugs occur somewhere between chair and a monitor.

**Comment by glenn-jocher** [comment]
@alienCarpet14 it's great to hear that you were able to identify the source of the issue! It's not uncommon for these kinds of challenges to arise during the development and deployment of models. Thank you for sharing your experience, and we hope your insights will be helpful to others encountering similar issues. If you have any further questions or need assistance, please feel free to reach out.
